rust/src/fvm/machine.rs

```diff
@@ -0,0 +1,264 @@
+        match machine {
+            Some(machine) => {
+                response.status_code = FCPResponseStatus::FCPNoError;
+            }
+            None => {
+                response.status_code = FCPResponseStatus::FCPUnclassifiedError;
+                response.error_msg = rust_str_to_c_str(format!("invalid machine id"));
+            }
+        }
+
+        info!("fil_drop_fvm_machine: end");
+
+        raw_ptr(response)
+    })
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn fil_fvm_machine_execute_message(
+    machine_id: u64,
+    message: fil_Message,
+    apply_kind: u64, /* 0: Explicit, _: Implicit */
+) -> *mut fil_FvmMachineExecuteResponse {
+    catch_panic_response(|| {
+        init_log();
+
+        info!("fil_fvm_machine_execute_message: start");
+
+        let mut response = fil_FvmMachineExecuteResponse::default();
+
+        let apply_kind = if apply_kind == 0 {
+            ApplyKind::Explicit
+        } else {
+            ApplyKind::Implicit
+        };
+
+        let message = match convert_fil_message_to_message(message) {
+            Ok(x) => x,
+            Err(err) => {
+                response.status_code = FCPResponseStatus::FCPUnclassifiedError;
+                response.error_msg = rust_str_to_c_str(format!("{:?}", err));
+                return raw_ptr(response);
+            }
+        };
+
+        let mut machines = FVM_MAP.lock().unwrap();
+        let mut machine = machines.get_mut(&machine_id);
+        match machine {
+            Some(machine) => {
+                let apply_ret = match machine.execute_message(message, apply_kind) {
+                    Ok(x) => x,
+                    Err(err) => {
+                        response.status_code = FCPResponseStatus::FCPUnclassifiedError;
+                        response.error_msg = rust_str_to_c_str(format!("{:?}", err));
+                        return raw_ptr(response);
+                    }
+                };
+
+                response.status_code = FCPResponseStatus::FCPNoError;
+                // FIXME: Return relevant fields of ApplyRet
+            }
+            None => {
+                response.status_code = FCPResponseStatus::FCPUnclassifiedError;
+                response.error_msg = rust_str_to_c_str(format!("invalid machine id"));
+            }
+        }
+
+        info!("fil_fvm_machine_execute_message: end");
+
+        raw_ptr(response)
+    })
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn fil_fvm_machine_finish_message(
+    machine_id: u64,
+    // TODO: actual message
+) {
+    // catch_panic_response(|| {
+    init_log();
+
+    info!("fil_fvm_machine_flush_message: start");
+
+    let machines = FVM_MAP.lock().unwrap();
+    let machine = machines.get(&machine_id);
+    match machine {
+        Some(machine) => {
+            todo!("execute message")
+        }
+        None => {
+            todo!("invalid machine id")
+        }
+    }
+
+    info!("fil_fvm_machine_flush_message: end");
+    // })
+}
```

Comment:
I don't think we have a "finish message" operation. What we finish is the _machine_, which flushes all effectively linked blocks to the blockstore, and returns the new state root CID.

@Stebalien do we expect the node to construct the receipts root, or do we want to accumulate receipts in the machine, to then return the root?

Reply:
I expect the node to deal with that, honestly. Basically:

1. The VM handles the state-tree _only_.
2. The Node handles everything related to the blockchain.